test/components/structures/TimelinePanel-test.js

```diff
@@ -276,16 +284,38 @@ describe('TimelinePanel', function() {
             // back-paginate until we hit the start
             return backPaginate();
         }).then(() => {
+            // hopefully, we got to the start of the timeline
+            expect(messagePanel.props.backPaginating).toBe(false);
+
             expect(messagePanel.props.suppressFirstDateSeparator).toBe(false);
             var events = scryEventTiles(panel);
-            expect(events[0].props.mxEvent).toBe(timeline.getEvents()[0])
+            expect(events[0].props.mxEvent).toBe(timeline.getEvents()[0]);
+            expect(events.length).toEqual(TIMELINE_CAP);
 
             // we should now be able to scroll down, and paginate in the other
             // direction.
             setScrollTop(scrollingDiv.scrollHeight);
             scrollingDiv.scrollTop = scrollingDiv.scrollHeight;
-            return awaitScroll();
+
+            // the delay() below is a heinous hack to deal with the fact that,
+            // without it, we may or may not get control back before the
+            // forward pagination completes. The delay means that it should
+            // have completed.
+            return awaitScroll().delay(0);
         }).then(() => {
+            expect(messagePanel.props.backPaginating).toBe(false);
+            expect(messagePanel.props.forwardPaginating).toBe(false);
+            expect(messagePanel.props.suppressFirstDateSeparator).toBe(true);
+
+            var events = scryEventTiles(panel);
+            expect(events.length).toEqual(TIMELINE_CAP);
+
+            // we don't really know what the first event tile will be, since that
+            // depends on how much the timelinepanel decides to paginate.
+            //
+            // just check that the first tile isn't event 0.
+            expect(events[0].props.mxEvent).toNotBe(timeline.getEvents()[0]);
+
             console.log("done");
         }).done(done, done);
```

Comment:
If we randomly throw, this will execute the rejected handler which in this case is `done()` which will complete the test. Will this actually *fail* the test, given there is no `expect()`?